pkg/driver/controller.go

```diff
@@ -25,28 +25,279 @@ import (
+			fsOptions.CopyTagsToBackups = boolVal
+		}
+		if val, ok := volumeParams[volumeParamsCopyTagsToVolumes]; ok {
+			boolVal, err := util.StringPointerToBoolPointer(&val)
+			if err != nil {
+				return nil, fmt.Errorf("invalid input for %s: %s", volumeParamsCopyTagsToVolumes, err)
+			}
+			fsOptions.CopyTagsToVolumes = boolVal
+		}
+		if val, ok := volumeParams[volumeParamsDailyAutomaticBackupStartTime]; ok {
+			fsOptions.DailyAutomaticBackupStartTime = &val
+		}
+		if val, ok := volumeParams[volumeParamsDeploymentType]; ok {
+			fsOptions.DeploymentType = &val
+		}
+		if val, ok := volumeParams[volumeParamsDiskIopsConfiguration]; ok {
+			fsOptions.DiskIopsConfiguration = &val
+		}
+		if val, ok := volumeParams[volumeParamsRootVolumeConfiguration]; ok {
+			fsOptions.RootVolumeConfiguration = &val
+		}
+		if val, ok := volumeParams[volumeParamsThroughputCapacity]; ok {
+			i, err := util.StringPointerToIntPointer(&val)
+			if err != nil {
+				return nil, fmt.Errorf("invalid input for %s: %s", volumeParamsThroughputCapacity, err)
+			}
+			fsOptions.ThroughputCapacity = i
+		}
+		if val, ok := volumeParams[volumeParamsWeeklyMaintenanceStartTime]; ok {
+			fsOptions.WeeklyMaintenanceStartTime = &val
+		}
+		if val, ok := volumeParams[volumeParamsSecurityGroupIds]; ok {
+			fsOptions.SecurityGroupIds = util.SplitCommasAndRemoveOuterBrackets(val)
+		}
+		if val, ok := volumeParams[volumeParamsSubnetIds]; ok {
+			fsOptions.SubnetIds = util.SplitCommasAndRemoveOuterBrackets(val)
+		}
+		if val, ok := volumeParams[volumeParamsTags]; ok {
+			fsOptions.Tags = &val
+		}
+
+		fs, err := d.cloud.CreateFileSystem(ctx, volName, fsOptions)
+		if err != nil {
+			klog.V(4).Infof("CreateFileSystem error: ", err.Error())
+			switch err {
+			case cloud.ErrInvalidParameter:
+				return nil, status.Error(codes.InvalidArgument, err.Error())
+			case cloud.ErrAlreadyExists:
+				return nil, status.Error(codes.AlreadyExists, err.Error())
+			default:
+				return nil, status.Errorf(codes.Internal, "Could not create volume %q: %v", volName, err)
+			}
+		}
+
+		err = d.cloud.WaitForFileSystemAvailable(ctx, fs.FileSystemId)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Filesystem is not ready: %v", err)
+		}
+
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				CapacityBytes: util.GiBToBytes(fs.StorageCapacity),
+				VolumeId:      fs.FileSystemId,
+				VolumeContext: map[string]string{
+					volumeContextDnsName:    fs.DnsName,
+					volumeContextVolumeType: volumeType,
+				},
+			},
+		}, nil
+	}
+	if volumeType == "volume" {
+		var storageCapacity *int64
+
+		if req.GetCapacityRange() != nil {
+			convertedCapacity := util.BytesToGiB(req.GetCapacityRange().GetRequiredBytes())
+			storageCapacity = &convertedCapacity
+		}
+
+		volOptions := cloud.VolumeOptions{
+			StorageCapacityQuotaGiB:       storageCapacity,
+			StorageCapacityReservationGiB: storageCapacity,
+		}
+
+		if val, ok := volumeParams[volumeParamsName]; ok {
```

Comment:
I messed this up in the design, but the storage class shouldn't have a "name" parameter. This will be pulled from the PVC, so you can set volOptions.Name=req.GetName()